src/com/ms/silverking/thread/lwt/LWTPoolImpl.java

```diff
@@ -14,266 +15,268 @@
-     * attempt to find an idle thread, or - failing that - a thread 
-     * with a small queue.
-     * 
-     * @param worker
-     * @param item
-     */
-    public void addWork(BaseWorker worker, Object item) {
-        addWork(worker, item, 0);
+    active = numActiveThreads();
+    loadStats.addLoadSample(active);
+    if (active < targetSize) {
+      increaseThreadsToLimit();
+    } else if (active > targetSize) {
+      decreaseThreadsToLimit();
     }
-    
-    /**
-     * Add work to be done and the worker that is to perform this work to 
-     * this pool. 
-     * 
-     * Direct call optimizations must have been performed prior to this
-     * call as this call always results in queueing.
-     * 
-     * If a common queue is in place, simply add. Otherwise, make an 
-     * attempt to find an idle thread, or - failing that - a thread 
-     * with a small queue.
-     * 
-     * @param worker
-     * @param item
-     */
-    public void addWork(BaseWorker worker, Object item, int priority) {
-        if (commonQueue != null) {
-            try {
-                commonQueue.put(new AssignedWork(worker, item, priority));
-            } catch (InterruptedException ie) {
-            }
-        } else {        
+  }
 
-            LWTThread   thread;
-            int         index;
-            
-            index = Math.abs(nextThread.getAndIncrement() % activeThreads.size());
-            thread = activeThreads.get(index);
-            //thread.addWork(worker, item);
-            
-            //throw new RuntimeException("deprecated");
-            // we would need to make thread list threadsafe
-            // if we wanted to get this
+  public int numIdleThreads() {
+    return idleThreads.size();
+  }
+
+  /**
+   * Add work to be done and the worker that is to perform this work to
+   * this pool.
+   * <p>
+   * Direct call optimizations must have been performed prior to this
+   * call as this call always results in queueing.
+   * <p>
+   * If a common queue is in place, simply add. Otherwise, make an
+   * attempt to find an idle thread, or - failing that - a thread
+   * with a small queue.
+   *
+   * @param worker
+   * @param item
+   */
+  public void addWork(BaseWorker worker, Object item) {
+    addWork(worker, item, 0);
+  }
+
+  /**
+   * Add work to be done and the worker that is to perform this work to
+   * this pool.
+   * <p>
+   * Direct call optimizations must have been performed prior to this
+   * call as this call always results in queueing.
+   * <p>
+   * If a common queue is in place, simply add. Otherwise, make an
+   * attempt to find an idle thread, or - failing that - a thread
+   * with a small queue.
+   *
+   * @param worker
+   * @param item
+   */
+  public void addWork(BaseWorker worker, Object item, int priority) {
+    if (commonQueue != null) {
+      try {
+        commonQueue.put(new AssignedWork(worker, item, priority));
+      } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)